File: client-apps/cli/internal/cli/temporal/supervisor.go
package temporal

import (
	"context"
	"time"

	"github.com/rs/zerolog/log"
)

const (
	// DefaultHealthCheckInterval is how often to check Temporal health
	DefaultHealthCheckInterval = 5 * time.Second
	
	// RestartBackoffDelay is the delay before attempting restart after failure
	RestartBackoffDelay = 1 * time.Second
)

// Supervisor monitors Temporal health and automatically restarts on failure
type Supervisor struct {
	manager           *Manager
	ctx               context.Context
	cancel            context.CancelFunc
	healthCheckInterval time.Duration
}

// NewSupervisor creates a new Temporal supervisor
func NewSupervisor(manager *Manager) *Supervisor {
	ctx, cancel := context.WithCancel(context.Background())
	
	return &Supervisor{
		manager:           manager,
		ctx:               ctx,
		cancel:            cancel,
		healthCheckInterval: DefaultHealthCheckInterval,
	}
}

// Start begins monitoring Temporal and auto-restarting on failure
// This function spawns a goroutine and returns immediately.
// The goroutine will run until Stop() is called or context is cancelled.
func (s *Supervisor) Start() {
	log.Info().
		Dur("interval", s.healthCheckInterval).
		Msg("Starting Temporal supervisor")
	
	go s.run()
}

// Stop gracefully stops the supervisor
func (s *Supervisor) Stop() {
	log.Info().Msg("Stopping Temporal supervisor")
	s.cancel()
}

// run is the main supervisor loop
func (s *Supervisor) run() {
	ticker := time.NewTicker(s.healthCheckInterval)
	defer ticker.Stop()
	
	for {
		select {
		case <-s.ctx.Done():
			log.Debug().Msg("Supervisor context cancelled, stopping")
			return
			
		case <-ticker.C:
			s.checkHealthAndRestart()
		}
	}
}

// checkHealthAndRestart checks Temporal health and restarts if unhealthy
func (s *Supervisor) checkHealthAndRestart() {
	// Check if Temporal is running and healthy
	if s.manager.IsRunning() {
		log.Debug().Msg("Temporal health check passed")
		return
	}
	
	// Temporal is not running or unhealthy - attempt restart
	log.Warn().Msg("Temporal health check failed - process not running or unhealthy")
	
	// Small backoff before restart attempt
	time.Sleep(RestartBackoffDelay)
	
	// Check context before attempting restart (supervisor may have been stopped)
	select {
	case <-s.ctx.Done():
		log.Debug().Msg("Supervisor stopped during restart backoff, aborting restart")
		return
	default:
	}
	
	log.Info().Msg("Attempting to restart Temporal...")
	
	// Use the idempotent Start() method which handles:
	// - Cleanup of stale processes
	// - Reusing healthy instances
	// - Starting new instance if needed
	if err := s.manager.Start(); err != nil {
		log.Error().
			Err(err).
			Msg("Failed to restart Temporal - will retry on next health check")
		return
	}
	
	log.Info().Msg("Temporal restarted successfully")
}

File: client-apps/cli/src/local/temporal/supervisor.ts
// Self-healing supervisor for the Temporal dev server.
//
// On a fixed interval it asks "is Temporal healthy?" and, if not, restarts it
// via the idempotent start path (which reuses healthy instances and cleans up
// stale ones). This is intentionally separate from the daemon's child
// supervisor: Temporal owns its own lifecycle and lock, so it carries its own
// watchdog — exactly the split the Go CLI uses.
//
// The check and backoff are dependency-injected (clock) so the policy is unit-
// testable without real time, and the loop never overlaps with itself.

import { type Logger, log as defaultLog } from "../../logger.js";

const DEFAULT_INTERVAL_MS = 5_000;
const DEFAULT_BACKOFF_MS = 1_000;

/** The minimal surface the supervisor drives. TemporalManager satisfies it. */
export interface SupervisedTarget {
  isRunning(): Promise<boolean>;
  start(): Promise<void>;
}

export interface SupervisorOptions {
  intervalMs?: number;
  backoffMs?: number;
  /** Injectable sleep, for deterministic tests. */
  sleep?: (ms: number) => Promise<void>;
  log?: Logger;
}

export class TemporalSupervisor {
  private readonly target: SupervisedTarget;
  private readonly intervalMs: number;
  private readonly backoffMs: number;
  private readonly sleep: (ms: number) => Promise<void>;
  private readonly log: Logger;
  private timer: NodeJS.Timeout | null = null;
  private stopped = false;
  private ticking = false;

  constructor(target: SupervisedTarget, options: SupervisorOptions = {}) {
    this.target = target;
    this.intervalMs = options.intervalMs ?? DEFAULT_INTERVAL_MS;
    this.backoffMs = options.backoffMs ?? DEFAULT_BACKOFF_MS;
    this.sleep = options.sleep ?? ((ms) => new Promise((r) => setTimeout(r, ms)));
    this.log = options.log ?? defaultLog;
  }

  /** Begin monitoring. Returns immediately; the loop runs on a timer. */
  start(): void {
    if (this.timer !== null) return;
    this.stopped = false;
    this.log.info("starting Temporal supervisor", { interval_ms: this.intervalMs });
    const tick = (): void => {
      void this.checkHealthAndRestart().finally(() => {
        if (!this.stopped) this.timer = setTimeout(tick, this.intervalMs);
      });
    };
    this.timer = setTimeout(tick, this.intervalMs);
  }

  /** Stop monitoring. Safe to call multiple times. */
  stop(): void {
    this.stopped = true;
    if (this.timer !== null) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    this.log.info("stopping Temporal supervisor");
  }

  /**
   * One health check + conditional restart. Public for unit testing: it is the
   * whole supervision policy, exercised without real timers.
   */
  async checkHealthAndRestart(): Promise<void> {
    if (this.ticking) return; // never overlap a previous tick
    this.ticking = true;
    try {
      if (await this.target.isRunning()) {
        this.log.debug("Temporal health check passed");
        return;
      }
      this.log.warn("Temporal health check failed — attempting restart");
      await this.sleep(this.backoffMs);
      if (this.stopped) return; // bailed out during backoff
      try {
        await this.target.start();
        this.log.info("Temporal restarted");
      } catch (err) {
        this.log.error("failed to restart Temporal — will retry next check", { error: String(err) });
      }
    } finally {
      this.ticking = false;
    }
  }
}
